soco/discovery.py

```diff
@@ -215,27 +243,180 @@ def any_soco():
             d for d in cls._instances[cls._class_group].values() if d.is_visible
         )
     except (KeyError, StopIteration):
-        devices = discover()
+        devices = discover(
+            allow_network_scan=allow_network_scan, min_netmask=min_netmask
+        )
         return None if devices is None else devices.pop()
 
     return device
 
 
-def by_name(name):
+def by_name(name, allow_network_scan=False, min_netmask=24):
     """Return a device by name.
 
     Args:
         name (str): The name of the device to return.
+        allow_network_scan (bool, optional): If normal discovery fails, fall
+            back to a scan of the attached network(s) to detect Sonos
+            devices.
+        min_netmask (int, optional): The minimum number of netmask bits. Used to
+            constrain the network search space when network scanning is
+            used.
 
     Returns:
         :class:`~.SoCo`: The first device encountered among all zone with the
             given player name. If none are found `None` is returned.
     """
-    devices = discover()
+    devices = discover(allow_network_scan=allow_network_scan, min_netmask=min_netmask)
     if devices is None:
         return None
 
     for device in devices:
         if device.player_name == name:
             return device
     return None
+
+
+def scan_network(max_threads=256, timeout=3.0, min_netmask=24, include_invisible=False):
```

Comment:
I was thinking a little about what a sensible default value of timeout would be here. Since that timeout applies to every individual network wait-for-reply, the max time the function takes is only capped at: `num_ip_addresses_in_network / max_threads * timeout`, so if the network is large, that time could be large and the real time it takes is at least proportional to the timeout. So the thing is that if we do get a reply, I would expect it to be fairly quick. On my own network this function return the correct result with timeout as low as 1E-5:
```
>>> scan_network(timeout=0.001)
{SoCo("192.168.1.173"), SoCo("192.168.1.197"), SoCo("192.168.1.31"), SoCo("192.168.1.18"), SoCo("192.168.1.215")}
>>> scan_network(timeout=1E-5)
{SoCo("192.168.1.173"), SoCo("192.168.1.197"), SoCo("192.168.1.31"), SoCo("192.168.1.18"), SoCo("192.168.1.215")}
>>> scan_network(timeout=1E-6)
>>>
```

One should of course always be careful with extrapolating from just one case, but setting it to say 0.1 would still be 1E4 more than my lowest observed success case. I know that you tested this on multiple different networks, so I would be interested to hear what your thoughts are.

Reply:
You're right. Changed in https://github.com/SoCo/SoCo/commit/1b81d6fa1dfcd8018f2736230546443a3f785721.

There's a bit of history behind this, back from when I wanted an exhaustive list of Sonos devices on the network for SoCo-CLI, and there were some outlier responses on my networks (ARP delays?). The shorter timeout should be OK in the majority of cases, although I sometimes saw timeouts in excess of 3s.